pkg/cloudprovider/aws/instancetypes.go

```diff
@@ -0,0 +1,214 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aws
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/ec2/ec2iface"
+	"github.com/awslabs/karpenter/pkg/apis/provisioning/v1alpha1"
+	"github.com/awslabs/karpenter/pkg/cloudprovider"
+	"github.com/awslabs/karpenter/pkg/utils/functional"
+	"github.com/patrickmn/go-cache"
+)
+
+type InstanceTypeProvider struct {
+	ec2               ec2iface.EC2API
+	vpc               *VPCProvider
+	instanceTypeCache *cache.Cache
+}
+
+func NewInstanceTypeProvider(ec2 ec2iface.EC2API, vpcProvider *VPCProvider) *InstanceTypeProvider {
+	return &InstanceTypeProvider{
+		ec2:               ec2,
+		vpc:               vpcProvider,
+		instanceTypeCache: cache.New(CacheTTL, CacheCleanupInterval),
+	}
+}
+
+// Get instance types that are availble per availability zone
+func (p *InstanceTypeProvider) Get(ctx context.Context, clusterName string, constraints *cloudprovider.Constraints) (map[string][]*ec2.InstanceTypeInfo, error) {
+	locations := constraints.Zones
+	// If no zone constraints were specified, use all zones that the cluster spans
+	if len(locations) == 0 {
+		var err error
+		locations, err = p.vpc.GetZones(ctx, clusterName)
+		if err != nil {
+			return nil, err
+		}
+	}
+	// locations could be zone names and zone ids, so normalize them to all be zone names
+	locations, err := p.normalizeZones(ctx, locations)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
minor: seems worth breaking out into separate method .. it's slightly hard to follow right now because `locations` keeps getting re-assigned (also I'm not sure we really want to introduce the concept of 'location' given that location really seems to actually mean 'zone'?)